File: leetcode/c++/LCR_149_cong_shang_dao_xia_da_yin_er_cha_shu_lcof.cpp
type TreeNode struct {
      Val int
      Left *TreeNode
      Right *TreeNode
}

func decorateRecord(root *TreeNode) []int {
    var que []*TreeNode 
    var res []int
    if root != nil {
        que = append(que, root)
    } else {
        return res
    }
    for len(que) != 0 {
        tmp := que[0]
        que = que[1:]
        res = append(res, tmp.Val)
        if tmp.Left != nil {
            que = append(que, tmp.Left)
        }
        if tmp.Right != nil {
            que = append(que, tmp.Right)
        }
    }
    return res
}

File: leetcode/golang/LCR_149_cong_shang_dao_xia_da_yin_er_cha_shu_lcof.go
// https://leetcode.cn/problems/cong-shang-dao-xia-da-yin-er-cha-shu-lcof/description

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> decorateRecord(TreeNode* root) {
        queue<TreeNode*> st;
        vector<int> res;
        if (root != nullptr)
            st.push(root);
        else 
            return {};
        while (!st.empty()) {
            TreeNode* node = st.front();
            st.pop();
            res.push_back(node->val);
            if (node->left != nullptr)
                st.push(node->left);
            if (node->right != nullptr)
                st.push(node->right);
        }
        return res;
    }
};

// 层次遍历，O(n)，O(n)